File: src/stores/workflow.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { useTauriIPC } from "../composables/useTauriIPC";

export interface WorkflowStage {
  name: string;
  stage_type: string;
  agent_id?: string;
  model_provider_id?: string;
  system_prompt?: string;
  temperature: number;
  max_tokens: number;
}

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  stages: WorkflowStage[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface WorkflowExecution {
  id: string;
  workflow_id: string;
  chapter_id: string;
  status: string;
  current_stage: number;
  results: Record<string, string>;
  error?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}

export const useWorkflowStore = defineStore("workflow", () => {
  const { call } = useTauriIPC();
  const workflows = ref<Workflow[]>([]);
  const currentExecution = ref<WorkflowExecution | null>(null);
  const loading = ref(false);

  const fetchWorkflows = async () => {
    loading.value = true;
    try {
      workflows.value = await call<Workflow[]>("list_workflows");
    } finally {
      loading.value = false;
    }
  };

  const createWorkflow = async (
    name: string,
    description: string,
    stages: WorkflowStage[],
  ) => {
    const workflow = await call<Workflow>("create_workflow", {
      name,
      description,
      stages,
    });
    workflows.value.unshift(workflow);
    return workflow;
  };

  const deleteWorkflow = async (id: string) => {
    await call("delete_workflow", { id });
    workflows.value = workflows.value.filter((w) => w.id !== id);
  };

  const executeWorkflow = async (workflowId: string, chapterId: string) => {
    const execution = await call<WorkflowExecution>("execute_workflow", {
      workflowId,
      chapterId,
    });
    currentExecution.value = execution;
    return execution;
  };

  return {
    workflows,
    currentExecution,
    loading,
    fetchWorkflows,
    createWorkflow,
    deleteWorkflow,
    executeWorkflow,
  };
});

File: src-tauri/src/models/workflow.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Workflow {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub stages: Vec<WorkflowStage>,
    pub is_active: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStage {
    pub name: String,
    pub stage_type: String, // outline/draft/proofread/edit/custom
    pub agent_id: Option<String>,
    pub model_provider_id: Option<String>,
    pub system_prompt: Option<String>,
    pub temperature: f64,
    pub max_tokens: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowExecution {
    pub id: String,
    pub workflow_id: String,
    pub chapter_id: String,
    pub status: String, // pending/running/completed/failed
    pub current_stage: usize,
    pub results: std::collections::HashMap<String, String>,
    pub error: Option<String>,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
    pub created_at: String,
}
